src/app/test_executive/network.ml

```diff
@@ -0,0 +1,140 @@
+open Core
+open Async
+
+type t =
+  { cluster: string
+  ; namespace: string
+  ; keypair_secrets: string list
+  ; testnet_dir: string
+  ; mutable deployed: bool }
+
+let run_cmd' testnet_dir prog args =
+  Process.create_exn ~working_dir:testnet_dir ~prog ~args ()
+  >>= Process.collect_output_and_wait
+
+let run_cmd_exn' testnet_dir prog args =
+  let open Process.Output in
+  let%bind output = run_cmd' testnet_dir prog args in
+  let print_output () =
+    let indent str =
+      String.split str ~on:'\n'
+      |> List.map ~f:(fun s -> "    " ^ s)
+      |> String.concat ~sep:"\n"
+    in
+    print_endline "=== COMMAND ===" ;
+    print_endline
+      (indent
+         ( prog ^ " "
+         ^ String.concat ~sep:" "
+             (List.map args ~f:(fun arg -> "\"" ^ arg ^ "\"")) )) ;
+    print_endline "=== STDOUT ===" ;
+    print_endline (indent output.stdout) ;
+    print_endline "=== STDERR ===" ;
+    print_endline (indent output.stderr) ;
+    Writer.(flushed (Lazy.force stdout))
+  in
+  match output.exit_status with
+  | Ok () ->
+      return ()
+  | Error (`Exit_non_zero status) ->
+      let%map () = print_output () in
+      failwithf "command exited with status code %d" status ()
+  | Error (`Signal signal) ->
+      let%map () = print_output () in
+      failwithf "command exited prematurely due to signal %d"
+        (Signal.to_system_int signal)
+        ()
+
+let run_cmd t prog args = run_cmd' t.testnet_dir prog args
+
+let run_cmd_exn t prog args = run_cmd_exn' t.testnet_dir prog args
+
+let create ~coda_automation_location ~testnet_name ~network_config =
+  let keypairs, direct_network_config =
+    Network_config.Abstract.render testnet_name network_config
+  in
+  let testnet_dir =
+    coda_automation_location ^/ "terraform/testnets" ^/ testnet_name
+  in
+  (* cleanup old deployment, if it exists; we will need to take good care of this logic when we put this in CI *)
+  let%bind () =
+    if%bind File_system.dir_exists testnet_dir then
+      let%bind () = run_cmd_exn' testnet_dir "terraform" ["refresh"] in
+      let%bind () =
+        let open Process.Output in
+        let%bind state_output =
+          run_cmd' testnet_dir "terraform" ["state"; "list"]
+        in
+        if not (String.is_empty state_output.stdout) then
+          run_cmd_exn' testnet_dir "terraform" ["destroy"; "-auto-approve"]
+        else return ()
+      in
+      File_system.remove_dir testnet_dir
+    else return ()
+  in
+  let%bind () = Unix.mkdir testnet_dir in
+  (* TODO: prebuild genesis proof and ledger *)
+  (*
```

Comment:
I think we should have the runtime_config defined outside and passed as input to the executive. That way, we can generate base proof and ledger as a separate CI Step and the nodes would then look for it in s3

Reply:
Yeah, that's a much better idea. Thanks.

Reply:
Hmm, actually, thinking about this some more...

I like the idea of caching proofs and ledgers in CI, and definitely think we should do that. But I'm less certain now about having the runtime config be entirely passed into the integration tests. If we pass it in from the outside, we may end up in the same situation we do today with integration tests, where we have to maintain a separate list of configs along with a script `test.py` which maps configs to tests. The main issue of this is that there ends up being a tight coupling between some of the tests and their configs (for instance, not all integration tests will pass on all configs). It would be nice if we could avoid this by defining the configurations inline with the tests. We can do something similar to what testground does for coupling multiple configurations to tests if we want to have different sizes or speeds of tests we run.